src/backend/distributed/operations/repair_shards.c

```diff
@@ -287,7 +419,57 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
 
 
 /*
- * master_move_shard_placement is a wrapper function for old UDF name.
+ * EnsureEnoughDiskSpaceForShardMove checks that there is enough space for
+ * shard moves of the given colocated shard list from source node to target node.
+ * It tries to clean up old shard placements to ensure there is enough space.
+ */
+static void
+EnsureEnoughDiskSpaceForShardMove(List *colocatedShardList,
+								  char *sourceNodeName, uint32 sourceNodePort,
+								  char *targetNodeName, uint32 targetNodePort)
+{
+	if (!CheckAvailableSpaceBeforeMove)
+	{
+		return;
+	}
+	uint64 colocationSizeInBytes = ShardListSizeInBytes(colocatedShardList,
+														sourceNodeName,
+														sourceNodePort);
+
+	uint32 connectionFlag = 0;
+	MultiConnection *connection = GetNodeConnection(connectionFlag, targetNodeName,
+													targetNodePort);
+	DeferredErrorMessage *temporaryError =
+		CheckSpaceConstraints(connection, colocationSizeInBytes);
+	if (temporaryError == NULL)
+	{
+		return;
+	}
+	RaiseDeferredError(temporaryError, WARNING);
+
+	/*
+	 * Only run shard cleanup when the first round of checks has failed
+	 * otherwise simply let the automatic cleanup do its job.
+	 *
+	 * NOTE: this tries to drop all marked shards, not only on the shard
+	 * that we're trying to move to.
+	 */
+	ereport(NOTICE, (errmsg("trying to drop old shards to resolve issue...")));
+
+	bool waitForCleanupLock = true;
+	DropMarkedShards(waitForCleanupLock);
```

Comment:
maybe check the return value of `DropMarkedShards` and if it is zero, print an error and return early. This is to give a more information to the user

Reply:
Also added information added how many shards are dropped